### ISSUE TITLE
Getting Shifted values of Close Price in Tensorflow training, validation and Testing. Python

### ISSUE THREAD

**Issue by JafferWilson** [tensorflow/tensorflow#21755]
I am trying to experiment with the Tensorflow. I am supplying the sequence of 5 out of which 4 are input and the 5th value is the output from the close price.
The graphs are plotted according to the respective values of the output as the original and the predicted for the training, validation and Testing.

But the graph shows shifted output. See the graphs:
![file](https://user-images.githubusercontent.com/13446197/44395357-44371b00-a557-11e8-8e22-94c4484fcbac.jpg)

I do not understand what is the problem as I am trying to get the prediction one step ahead of the given input.

Here is the training code that I am trying to run:

```
index_in_epoch = 0;
perm_array  = np.arange(x_train.shape[0])
np.random.shuffle(perm_array)

# function to get the next batch
def get_next_batch(batch_size):
    global index_in_epoch, x_train, perm_array   
    start = index_in_epoch
    index_in_epoch += batch_size

    if index_in_epoch > x_train.shape[0]:
        np.random.shuffle(perm_array) # shuffle permutation array
        start = 0 # start next epoch
        index_in_epoch = batch_size

    end = index_in_epoch
    return x_train[perm_array[start:end]], y_train[perm_array[start:end]]

# parameters
n_steps = seq_len-1 
n_inputs = x_train.shape[2]#4
n_neurons = 500
n_outputs = y_train.shape[1]#4
n_layers = 2
learning_rate = 0.0001
batch_size =10
n_epochs = 1000#200 
train_set_size = x_train.shape[0]
test_set_size = x_test.shape[0]

tf.reset_default_graph()

X = tf.placeholder(tf.float32, [None, n_steps, n_inputs])
y = tf.placeholder(tf.float32, [None, n_outputs])

layers = [tf.contrib.rnn.LSTMCell(num_units=n_neurons, 
                                 activation=tf.nn.leaky_relu, use_peepholes = True)
         for layer in range(n_layers)]

multi_layer_cell = tf.contrib.rnn.MultiRNNCell(layers)
rnn_outputs, states = tf.nn.dynamic_rnn(multi_layer_cell, X, dtype=tf.float32)

stacked_rnn_outputs = tf.reshape(rnn_outputs, [-1, n_neurons]) 
stacked_outputs = tf.layers.dense(stacked_rnn_outputs, n_outputs)
outputs = tf.reshape(stacked_outputs, [-1, n_steps, n_outputs])
outputs = outputs[:,n_steps-1,:] # keep only last output of sequence

loss = tf.reduce_mean(tf.square(outputs - y)) # loss function = mean squared error 
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate) 
training_op = optimizer.minimize(loss)

saver = tf.train.Saver()
display = 40
with tf.Session() as sess: 
    sess.run(tf.global_variables_initializer())
    plt.ion()
    fig = plt.figure()
    fig.set_size_inches(30,10)
    ax1 = fig.add_subplot(231)
    line1, = ax1.plot(y_train[:display],color='blue', label='train original',marker=".")
    line2, = ax1.plot(y_train[:display],color='red', label='train Prediction',marker=".")
    ax2 = fig.add_subplot(232)
    line3, = ax2.plot(y_valid[:display],color='blue', label='valid original',marker=".")
    line4, = ax2.plot(y_valid[:display],color='red', label='valid Prediction',marker=".")
    ax3 = fig.add_subplot(233)
    line5, = ax3.plot(y_test[:display],color='blue', label='valid original',marker=".")
    line6, = ax3.plot(y_test[:display],color='red', label='valid Prediction',marker=".")
    ax4 = fig.add_subplot(234)
    candlestick2_ohlc(ax4,df_train_candle.o.values[:display],df_train_candle.h.values[:display],df_train_candle.l.values[:display],df_train_candle.c.values[:display],width=0.6)
    ax5 = fig.add_subplot(235)
    candlestick2_ohlc(ax5,df_valid_candle.o.values[:display],df_valid_candle.h.values[:display],df_valid_candle.l.values[:display],df_valid_candle.c.values[:display],width=0.6)
    ax6 = fig.add_subplot(236)
    candlestick2_ohlc(ax6,df_test_candle.o.values[:display],df_test_candle.h.values[:display],df_test_candle.l.values[:display],df_test_candle.c.values[:display],width=0.6)
    ax1.set_title('Training')
    ax2.set_title('Validation')
    ax3.set_title('Testing')
    ax1.set_xlabel('time')
    ax2.set_xlabel('time')
    ax3.set_xlabel('time')
    ax1.set_ylabel("Train Reversal")
    ax2.set_ylabel("Valid Reversal")
    ax3.set_ylabel("Test Reversal")

    ax4.set_title('Training Candles')
    ax4.set_xlabel('time')
    ax4.set_ylabel("Train OHLC")
    ax5.set_title('Validation Candles')
    ax5.set_xlabel('time')
    ax4.set_ylabel("Valid OHLC")
    ax6.set_title('Testing Candles')
    ax6.set_xlabel('time')
    ax4.set_ylabel("Test OHLC")


    plt.show()
    #if(tf.train.checkpoint_exists(tf.train.latest_checkpoint("modelsOHLC"))):
     #   saver.restore(sess, tf.train.latest_checkpoint("modelsOHLC"))
      #  print(tf.train.latest_checkpoint("modelsOHLC") + "Session Loaded for Testing")
    for iteration in range(int(n_epochs*train_set_size/batch_size)):
        x_batch, y_batch = get_next_batch(batch_size) # fetch the next training batch 
        sess.run(training_op, feed_dict={X: x_batch, y: y_batch}) 
        if iteration % int(1*train_set_size/batch_size) == 0:
            mse_train = loss.eval(feed_dict={X: x_train, y: y_train}) 
            mse_valid = loss.eval(feed_dict={X: x_valid, y: y_valid}) 
            mse_test = loss.eval(feed_dict={X: x_test, y: y_test})
            y_train_pred = sess.run(outputs, feed_dict={X: x_train})
            y_valid_pred = sess.run(outputs, feed_dict={X: x_valid})
            y_test_pred = sess.run(outputs, feed_dict={X: x_test})
            line2.set_ydata(y_train_pred[:display])
            line4.set_ydata(y_valid_pred[:display])
            line6.set_ydata(y_test_pred[:display])
            ax1.set_title("Training Loss: "+str(mse_train))
            ax2.set_title("Validation Loss: "+str(mse_valid))
            ax3.set_title("Testing Loss: "+str(mse_test))
            plt.pause(0.01)
            print('%.2f epochs: MSE train/valid/test = %.10f/%.10f/%.10f'%(
                iteration*batch_size/train_set_size, mse_train, mse_valid,mse_test))
            save_path = saver.save(sess, "modelsOHLC\\model"+str(iteration)+".ckpt")
```

The input data for the above is here

[Input file](https://gist.github.com/JafferWilson/c17a4d70b4aaf839b76bde13f7e32141)

You can access the Jupyter version of the code from here: [Jupyter Version of the Code
](https://gist.github.com/JafferWilson/2f7a2374e7b5ea4f92c2edda8b9f7691)
Please guys let me know what I have missed due to which the graph is getting shifted. I am in an ambiguous situation, please let me know what I can do to correct it.   

I have opened a question on Stackoverflow too, here is the [link for it](https://stackoverflow.com/questions/51945064/getting-shifted-values-of-close-price-in-tensorflow-training-validation-and-tes).

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by JafferWilson** [comment]
Here's what to be added. Please check the details.
Have I written custom code --- N/A
OS Platform and Distribution ---- Windows 10 
TensorFlow installed from --- pip installer
TensorFlow version ---- 1.9.0
Bazel version --- N/A
CUDA/cuDNN version ---- 9.0
GPU model and memory ----- I do not have the GPU mode right now. The system memory is around 32 GB and the GP memory is 12 GB. I am using Titan GTX -Titan Xp.
Exact command to reproduce --- This I have already provided, please check it in the question. Even you will get the full code with dataset that I tried.
Mobile device --- N/A

Let me know if anything else needed please.

**Comment by drpngx** [comment]
I think it's more of a stackoverflow question, since it's not really a bug or a feature request AFAICT. So I will close this one.

For best results, I would encourage you to ask a more targeted question -- regarding a specific problem that fits in two or three lines of code.

**Comment by JafferWilson** [comment]
@drpngx Thank you for your reply sir. But there isn't anyone answering to me on Stackoverflow. Hence, I thought this is the issue with Tensorflow itself. Please let me know if there is anything you can do to help me. I tried everything and this is my last attempt to resolution.  Please help me.

**Comment by drpngx** [comment]
Please try to find a small test case that exhibits the problem. It's hard to analyze a longer program.

**Comment by drpngx** [comment]
@JafferWilson Can you rephrase in a smaller case what the problem is? I don't understand what the shifting problem is.

**Comment by JafferWilson** [comment]
@drpngx Sure, I can. The graph in images you see:-- The blue color is the real data graph. The red one is predicted. You see the red one is the exact match of the blue graph. But the red graph is getting shifted one step ahead of the real one which is not correct. The output of the predicted must match the real values then only we can say that the prediction is perfect.
If they are not matching or getting plotted one step ahead of what is required, then there are some issue. I am not able to understand why tensorflow is predicting the values one extra step where as the real values are equal to what the predicted values are.
Can you please have a look at the issue? I have tried everything I could but am not able to understand the issue. My guess is that its an issue of the tensorflow itself.
Well, I guess no one is able to give solution to me. So I better stop thinking about it.

**Comment by drpngx** [comment]
Can you try to have `x[t] = 0.01 * xrange(100)`, and train/test on the same data? What happens then?

**Comment by JafferWilson** [comment]
@drpngx Sir, where exactly you want me to edit this part in my code? Please let me know I will try it out.

**Comment by hastirad** [comment]
@JafferWilson hi. I have same problem and my predictions had one lag. Could you solve the problem?

**Comment by JafferWilson** [comment]
@hastirad I don't know. that's why I ask a query here. If I had solution for this I would have already gone with Tensorflow. Currently, I am using a hybrid system with Tensorflow and other frameworks. But none is useful. Need to create my own framework.
